README.md
# barcode_info

A new Flutter package project.

## Getting Started

This project is a starting point for a Dart
[package](https://flutter.dev/developing-packages/),
a library module containing code that can be shared easily across
multiple Flutter or Dart projects.

For help getting started with Flutter, view our 
[online documentation](https://flutter.dev/docs), which offers tutorials, 
samples, guidance on mobile development, and a full API reference.


CHANGELOG.md

## 0.2.2 - 2023-03-06
### Changed
- ean13이 isbn13인 경우 자동으로 isbn으로 변경 처리

## 0.2.1 - 2023-03-06
### Changed
- barcode_widget: ^2.0.1 -> ^2.0.3
- tuple: ^2.0.0 -> ^2.0.1

### Added
- isbn13_info 추가

## 0.2.0 - 2021-06-19
### Changed
- null safety 지원

### Removed
- CountryInfo null safety적용문제로 제거, 기존 프로젝트는 개별사용하는 방식으로 변경 필요.

## 0.1.2 - 2020-12-13
### Added
- add BarcodeInfo.formatToString()

## 0.1.1 - 2020-12-11
### Fixed
- fix BarCodeInfo.fromString()

## 0.1.0 - 2020-12-09
### Removed
- remove asynchronous code

## 0.0.6 - 2020-12-09
### Added
- add BarcodeInfo.toString()
- add BarCodeInfo.fromString()

## 0.0.5 - 2020-12-02
### Added
- add BarcodeView widget

## 0.0.4 - 2020-11-29
### Added
- Support BarcodeFormat string for flutter_barcode_reader package.

## 0.0.3 - 2020-11-29
### Fixed
- bug fix

## 0.0.2 - 2020-11-29
### Added
- Lotto6/45 QRCode parser

## [0.0.1] - 2020-11-29
* TODO: Describe initial release.


test/barcode_info_test.dart
import 'dart:math';

import 'package:barcode_info/src/isbn13_info.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:barcode_info/barcode_info.dart';

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();
  test('로또6/45 QRCode parsing', () {
    const code =
        'http://m.dhlottery.co.kr/?v=0936m041720353943q071215303244m071113171829q010222253142m0414182731421050908394';
    final BarcodeInfo barcode = BarcodeInfo.create('QR_CODE', code);
    expect(barcode is Lotto645QrCodeInfo, true);
    final lottocode = barcode as Lotto645QrCodeInfo;
    expect(lottocode.drawNo, 936);
    expect(lottocode.games.length, 5);

    expect(lottocode.games[0].item1, 2);
    expect(lottocode.games[0].item2, equals([4, 17, 20, 35, 39, 43]));

    expect(lottocode.games[1].item1, 1);
    expect(lottocode.games[1].item2, equals([7, 12, 15, 30, 32, 44]));

    expect(lottocode.games[2].item1, 2);
    expect(lottocode.games[2].item2, equals([7, 11, 13, 17, 18, 29]));

    expect(lottocode.games[3].item1, 1);
    expect(lottocode.games[3].item2, equals([1, 2, 22, 25, 31, 42]));

    expect(lottocode.games[4].item1, 2);
    expect(lottocode.games[4].item2, equals([4, 14, 18, 27, 31, 42]));

    expect(lottocode.trNumber, 1050908394);
  });

  test('BarcodeInfo.toString() <-> BarcodeInfo.fromString() test', () {
    var code = '12344567';
    var barcode = BarcodeInfo.create('ean8', code);
    expect(barcode is Ean8Info, true);
    expect(barcode.toString(), 'ean8,12344567');
    var newBarcode = BarcodeInfo.fromString(barcode.toString());
    expect(newBarcode is Ean8Info, true);

    code =
        'http://m.dhlottery.co.kr/?v=0936m041720353943q071215303244m071113171829q010222253142m0414182731421050908394';
    barcode = BarcodeInfo.create('qrcode', code);
    expect(barcode.toString(),
        'qrCode,http://m.dhlottery.co.kr/?v=0936m041720353943q071215303244m071113171829q010222253142m0414182731421050908394');
    newBarcode = BarcodeInfo.fromString(barcode.toString());
    expect(newBarcode is Lotto645QrCodeInfo, true);
  });

  test('ISBN-13 test', () {
    var code = '9788175257665';
    final isISBN = EAN13Info.isISBN(code.substring(0, 3));
    expect(isISBN, true);

    var barcode = BarcodeInfo.create('isbn', code);
    expect(barcode is ISBN13Info, true);

    barcode = BarcodeInfo.create('ean13', code);
    expect(barcode is ISBN13Info, true);

    final isbn = barcode as ISBN13Info;
    expect(isbn.ean, 978);
    expect(isbn.group, 81);
    expect(isbn.publisher, 7525);
    expect(isbn.title, 766);
    expect(isbn.checkDigit, 5);
  });
}


lib/src/widgets.dart
import 'package:barcode_widget/barcode_widget.dart';
import 'package:flutter/material.dart';

import '../barcode_info.dart';

class BarcodeView extends StatelessWidget {
  final BarcodeInfo data;

  BarcodeView(this.data, {Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final barcode = _getBarcode();
    final barcodeWidget = BarcodeWidget(data: data.code, barcode: barcode);
    return Column(mainAxisAlignment: MainAxisAlignment.spaceBetween, children: [
      Spacer(),
      AspectRatio(
          aspectRatio: _getBarcodeAspectRatio(),
          child: Card(
              child: Padding(
                  padding: EdgeInsets.all(20),
                  child: Center(child: barcodeWidget)))),
      Spacer(),
    ]);
  }

  double _getBarcodeAspectRatio() {
    switch (data.format) {
      case BarcodeFormat.ean8:
        return 21.38 / 17.05;

      case BarcodeFormat.ean13:
        return 31.35 / 22.85;

      // 특별한 기준이 없는건 2:1로
      case BarcodeFormat.code39:
      case BarcodeFormat.code93:
      case BarcodeFormat.code128:
      case BarcodeFormat.codabar:
      case BarcodeFormat.itf:
        return 2 / 1;

      case BarcodeFormat.pdf417:
        return 2 / 1;

      case BarcodeFormat.aztec:
      case BarcodeFormat.dataMatrix:
      case BarcodeFormat.qrCode:
        return 1;

      case BarcodeFormat.upcA:
        return 37.29 / 25.91;

      case BarcodeFormat.upcE:
        return 22.11 / 25.91;

      default:
        return 1;
    }
  }

  Barcode _getBarcode() {
    switch (data.format) {
      case BarcodeFormat.aztec:
        return Barcode.aztec();

      case BarcodeFormat.codabar:
        return Barcode.codabar();

      case BarcodeFormat.code128:
        return Barcode.code128();

      case BarcodeFormat.code39:
        return Barcode.code39();

      case BarcodeFormat.code93:
        return Barcode.code93();

      case BarcodeFormat.dataMatrix:
        return Barcode.dataMatrix();

      case BarcodeFormat.ean8:
        return Barcode.ean8();

      case BarcodeFormat.ean13:
        return Barcode.ean13();

      case BarcodeFormat.itf:
        return Barcode.itf();

      case BarcodeFormat.pdf417:
        return Barcode.pdf417();

      case BarcodeFormat.qrCode:
        return Barcode.qrCode(
            errorCorrectLevel: BarcodeQRCorrectionLevel.quartile);

      case BarcodeFormat.upcA:
        return Barcode.upcA();

      case BarcodeFormat.upcE:
        return Barcode.upcE();

      default:
        // 지원하지 않는 포맷은 qrcode로 대체한다.
        return Barcode.qrCode(
            errorCorrectLevel: BarcodeQRCorrectionLevel.quartile);
    }
  }
}


lib/src/isbn13_info.dart
import 'barcode_base.dart';

class ISBN13Info extends BarcodeInfo {
  ISBN13Info(String code, this.ean, this.group, this.publisher, this.title,
      this.checkDigit)
      : super(BarcodeFormat.isbn13, code);

  factory ISBN13Info.fromCode(String code) {
    int ean = 0;
    int group = 0;
    int publisher = 0;
    int title = 0;
    int checkDigit = 0;
    if (code.length == 13) {
      ean = int.tryParse(code.substring(0, 3)) ?? 0;
      group = int.tryParse(code.substring(3, 5)) ?? 0;
      publisher = int.tryParse(code.substring(5, 9)) ?? 0;
      title = int.tryParse(code.substring(9, 12)) ?? 0;
      checkDigit = int.tryParse(code.substring(12, 13)) ?? 0;
    }
    return ISBN13Info(code, ean, group, publisher, title, checkDigit);
  }

  /// https://en.wikipedia.org/wiki/ISBN
  final int ean;
  final int group;
  final int publisher;
  final int title;
  final int checkDigit;
}


lib/src/qrcode_info.dart
import 'package:tuple/tuple.dart';

import 'barcode_base.dart';

class QrCodeInfo extends BarcodeInfo {
  QrCodeInfo(String code) : super(BarcodeFormat.qrCode, code);

  factory QrCodeInfo.fromCode(String code) {
    var info = Lotto645QrCodeInfo.tryParse(code);
    if (info != null) return info;
    return QrCodeInfo(code);
  }
}

/// Lotto6/45 (Korean lottery game) QrCode format
class Lotto645QrCodeInfo extends QrCodeInfo {
  /// 추첨회차
  final int drawNo;

  /// 구매 게임 목록
  ///
  /// ```
  ///   [
  ///     {번호선택유형, {선택번호 6개}},
  ///     {번호선택유형, {선택번호 6개}},
  ///     ...
  ///   ]
  /// ```
  /// 번호선택유형 : 0-없음, 1-자동, 2-수동, 3-반자동
  final List<Tuple2<int, Set<int>>> games;

  /// TR번호
  final int trNumber;

  Lotto645QrCodeInfo(String code, this.drawNo, this.games, this.trNumber)
      : super(code);

  /// QrCode값이 로또6/45코드인지 확인
  ///
  /// 형식
  /// 복권수탁사업자url/?v=[0000][m/s/q/n][000000000000][m/s/q/n][000000000000][m/s/q/n][000000000000][m/s/q/n][000000000000][m/s/q/n][000000000000][0000000000]
  static Lotto645QrCodeInfo? tryParse(final String code) {
    if (code.isEmpty) return null;

    // 로또6/45 수탁사업자 url여부 확인
    final baseUrls = [
      'http://qr.645lotto.net/?',
      'http://qr.nlotto.co.kr/?',
      'http://m.dhlottery.co.kr/?',
    ];

    var val = code;
    bool isLottoUrl = false;
    for (var baseUrl in baseUrls) {
      if (code.startsWith(baseUrl)) {
        isLottoUrl = true;
        val = code.substring(baseUrl.length);
        break;
      }
    }

    if (!isLottoUrl) return null;

    /// 데이터 길이 확인 (81자)
    /// http://m.dhlottery.co.kr/?v=0936m041720353943q071215303244m071113171829q010222253142m0414182731421050908394
    /// v=[0000][m/s/q/n][000000000000][m/s/q/n][000000000000][m/s/q/n][000000000000][m/s/q/n][000000000000][m/s/q/n][000000000000][0000000000]
    /// v= + 회차4자리 + (번호선택유형과 번호 13자리 * 5게임) + TR번호 10자리 = 총81자
    if (val.length != 81) return null;

    final drawNo = int.tryParse(val.substring(2, 6));
    if (drawNo == null) return null;
    val = val.substring(6);

    final games = <Tuple2<int, Set<int>>>[];
    for (var i = 0; i < 5; i++) {
      var s = val.substring(0, 13);
      val = val.substring(13);

      /// 번호선택유형
      var selectOption = 0;
      switch (s.substring(0, 1)) {
        case 'q':
          selectOption = 1;
          break;
        case 'm':
          selectOption = 2;
          break;
        case 's':
          selectOption = 3;
          break;
        default:
          continue;
      }

      s = s.substring(1);

      var lottoNumber = Set<int>();
      for (var j = 0; j < 6; j++) {
        final no = int.tryParse(s.substring(0, 2));
        if (no == null) return null;

        lottoNumber.add(no);
        s = s.substring(2);
      }

      games.add(Tuple2<int, Set<int>>(selectOption, lottoNumber));
    }

    final trNumber = int.tryParse(val);
    if (trNumber == null) return null;

    return Lotto645QrCodeInfo(code, drawNo, games, trNumber);
  }
}


lib/src/barcode_base.dart
import 'ean13_info.dart';
import 'isbn13_info.dart';
import 'qrcode_info.dart';
import 'upca_info.dart';

export 'ean13_info.dart';
export 'upca_info.dart';

enum BarcodeFormat {
  unknown,
  aztec,
  codabar,
  code128,
  code39,
  code93,
  dataMatrix,
  ean8,
  ean13,
  itf,
  maxicode,
  pdf417,
  qrCode,
  rss14,
  upcA,
  upcE,
  isbn13,
}

abstract class BarcodeInfo {
  BarcodeInfo(this.format, this.code);

  /// Create BarcodeInfo
  ///
  /// [format] format string (ignore case)
  /// * BarcodeFormat string list
  ///   - aztec       : 'AZTEC'
  ///   - codabar     : 'CODABAR'
  ///   - code128     : 'CODE_128' or 'CODE128'
  ///   - code39      : 'CODE_39' or 'CODE39'
  ///   - code93      : 'CODE_93' or 'CODE93'
  ///   - dataMatrix  : 'DATA_MATRIX' or 'DATAMATRIX'
  ///   - ean8        : 'EAN_8' or 'EAN8'
  ///   - ean13       : 'EAN_13' or 'EAN13'
  ///   - itf         : 'ITF' or 'INTERLEAVED2OF5'
  ///   - maxicode    : 'MAXICODE'
  ///   - pdf417      : 'PDF_417' or 'PDF417'
  ///   - qrCode      : 'QR_CODE' or 'QR' or 'QRCODE'
  ///   - rss14       : 'RSS_14' or 'RSS14'
  ///   - upcA        : 'UPC_A' or 'UPCA'
  ///   - upcE        : 'UPC_E' or 'UPCE'
  ///   - isbn        : 'ISBN' or 'ISBN13'
  static BarcodeInfo create(final String format, String code) {
    final fmt = format.toUpperCase();
    if (fmt == 'AZTEC') {
      return AxtecInfo(code);
    } else if (fmt == 'CODABAR') {
      return CodabarInfo(code);
    } else if (fmt == 'CODE_128' || fmt == 'CODE128') {
      return Code128Info(code);
    } else if (fmt == 'CODE_39' || fmt == 'CODE39') {
      return Code39Info(code);
    } else if (fmt == 'CODE_93' || fmt == 'CODE93') {
      return Code93Info(code);
    } else if (fmt == 'DATA_MATRIX' || fmt == 'DATAMATRIX') {
      return DataMatrixInfo(code);
    } else if (fmt == 'EAN_8' || fmt == 'EAN8') {
      return Ean8Info(code);
    } else if (fmt == 'EAN_13' || fmt == 'EAN13') {
      var ean13 = EAN13Info.fromCode(code);
      if (EAN13Info.isISBN(ean13.prefix)) {
        return ISBN13Info.fromCode(code);
      }
      return ean13;
    } else if (fmt == 'ITF' || fmt == 'INTERLEAVED2OF5') {
      return ItfInfo(code);
    } else if (fmt == 'MAXICODE') {
      return MaxicodeInfo(code);
    } else if (fmt == 'PDF_417' || fmt == 'PDF417') {
      return Pdf417Info(code);
    } else if (fmt == 'QRCODE' || fmt == 'QR_CODE' || fmt == 'QR') {
      return QrCodeInfo.fromCode(code);
    } else if (fmt == 'RSS_14' || fmt == 'RSS14') {
      return Rss14Info(code);
    } else if (fmt == 'UPC_A' || fmt == 'UPCA') {
      return UPCAInfo(code);
    } else if (fmt == 'UPC_E' || fmt == 'UPCE') {
      return UPCEInfo(code);
    } else if (fmt == 'ISBN' || fmt == 'ISBN13') {
      return ISBN13Info.fromCode(code);
    } else {
      return BarcodeUnknown(code, format);
    }

    /// [https://zxing.github.io/zxing/apidocs/com/google/zxing/BarcodeFormat.html]
    /// EAN_13, UPC_A, QR_CODE, DATA_MATRIX, AZTEC, CODABAR, CODE_128, CODE_39
    /// CODE_93, EAN_8, ITF, MAXICODE, PDF_417, RSS_14, UPC_E
    ///
    ///  RSS_EXPANDED, UPC_EAN_EXTENSION
  }

  /// string format => 'format,code'
  /// example:
  ///   'ean13,8801051262995'
  factory BarcodeInfo.fromString(String value) {
    final idx = value.indexOf(',');
    if (idx <= 0 || idx >= value.length - 1) {
      throw FormatException('unexpected format.');
    }

    final format = value.substring(0, idx);
    final code = value.substring(idx + 1);

    return create(format, code);
  }

  /// BarcodeFormat
  final BarcodeFormat format;

  /// Barcode value
  final String code;

  String formatToString() {
    return format.toString().split('.').last;
  }

  @override
  String toString() {
    return formatToString() + ',' + code;
  }
}

class BarcodeUnknown extends BarcodeInfo {
  BarcodeUnknown(String code, this.formatText)
      : super(BarcodeFormat.unknown, code);

  final String formatText;
}

class DataMatrixInfo extends BarcodeInfo {
  DataMatrixInfo(String code) : super(BarcodeFormat.dataMatrix, code);
}

class AxtecInfo extends BarcodeInfo {
  AxtecInfo(String code) : super(BarcodeFormat.aztec, code);
}

class CodabarInfo extends BarcodeInfo {
  CodabarInfo(String code) : super(BarcodeFormat.codabar, code);
}

class Code128Info extends BarcodeInfo {
  Code128Info(String code) : super(BarcodeFormat.code128, code);
}

class Code39Info extends BarcodeInfo {
  Code39Info(String code) : super(BarcodeFormat.code39, code);
}

class Code93Info extends BarcodeInfo {
  Code93Info(String code) : super(BarcodeFormat.code93, code);
}

class Ean8Info extends BarcodeInfo {
  Ean8Info(String code) : super(BarcodeFormat.ean8, code);
}

class ItfInfo extends BarcodeInfo {
  ItfInfo(String code) : super(BarcodeFormat.itf, code);
}

class MaxicodeInfo extends BarcodeInfo {
  MaxicodeInfo(String code) : super(BarcodeFormat.maxicode, code);
}

class Pdf417Info extends BarcodeInfo {
  Pdf417Info(String code) : super(BarcodeFormat.pdf417, code);
}

class Rss14Info extends BarcodeInfo {
  Rss14Info(String code) : super(BarcodeFormat.rss14, code);
}

class UPCEInfo extends BarcodeInfo {
  UPCEInfo(String code) : super(BarcodeFormat.upcE, code);
}


lib/src/ean13_info.dart
import 'barcode_base.dart';

class EAN13Info extends BarcodeInfo {
  EAN13Info(String code, this.prefix, {this.countryCodes})
      : super(BarcodeFormat.ean13, code);

  factory EAN13Info.fromCode(String code) {
    final prefix = code.substring(0, 3);
    final countryCodes = countriesFromGS1Prefix(prefix);

    return EAN13Info(code, prefix, countryCodes: countryCodes);
  }

  static bool isISBN(String prefix) {
    if (prefix.length != 3) return false;
    final code = int.tryParse(prefix);
    if (code == null) return false;
    // "Bookland" (ISBN) - 979-0 used for sheet music
    // ("Musicland", ISMN-13, replaces deprecated ISMN M-numbers)
    return (978 <= code && code <= 979);
  }

  /// GS1 prefix
  final String prefix;
  final Set<String>? countryCodes;
}

/// [GS1 Company Prefix](https://www.gs1.org/standards/id-keys/company-prefix)
/// https://en.wikipedia.org/wiki/List_of_GS1_country_codes
Set<String>? countriesFromGS1Prefix(final String prefix) {
  if (prefix.length != 3) return null;

  /// link : [https://en.wikipedia.org/wiki/List_of_GS1_country_codes]

  final code = int.tryParse(prefix);
  if (code == null) return null;

  /// description은 언어별로 다르게 표시해야하기 때문에 일단 링크만 설정하고
  /// 추후에 어떤값을 지정해야할지 고민한다.
  //String description =
  //    'https://en.wikipedia.org/wiki/List_of_GS1_country_codes';

  /// [https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2]
  /// 위 리스트에 없는 경우 국가명
  /// gs1 page와 wiki page의 정보가 약간 상이하다.
  /// 아래의 기준으로 작성됨.
  /// 1. wiki기준으로 작성
  /// 2. 국가가 2개 이상인 경우 gs1페이지 기준으로 1개 국가만 설정
  /// 3. gs1 page도 2개 국가 이상이면 그대로 설정
  final countries = Set<String>();

  if (0 <= code && code <= 19) {
    countries.addAll(['US']); //, 'CA']);
  } else if (20 <= code && code <= 29) {
    // Used to issue restricted circulation numbers within a geographic region
  } else if (30 <= code && code <= 39) {
    countries.add('US');
  } else if (40 <= code && code <= 49) {
    // Used to issue restricted circulation numbers within a geographic region
  } else if (50 <= code && code <= 59) {
    // GS1 US reserved for future use
  } else if (60 <= code && code <= 99) {
    countries.addAll(['US']); //, 'CA']);
  } else if (100 <= code && code <= 139) {
    countries.add('US');
  } else if (200 <= code && code <= 299) {
    // Used to issue restricted circulation numbers within a geographic region
  } else if (300 <= code && code <= 379) {
    countries.addAll(['FR']); //, 'MC']);
  } else if (code == 380) {
    countries.add('BG');
  } else if (code == 383) {
    countries.add('SI');
  } else if (code == 385) {
    countries.add('HR');
  } else if (code == 387) {
    countries.add('BA');
  } else if (code == 389) {
    countries.add('ME');
  } else if (code == 390) {
    countries.add('Kosovo');
  } else if (400 <= code && code <= 440) {
    countries.add('DE');
  } else if (450 <= code && code <= 459) {
    countries.add('JP');
  } else if (460 <= code && code <= 469) {
    countries.add('RU');
  } else if (code == 470) {
    countries.add('KG');
  } else if (code == 471) {
    countries.add('TW');
  } else if (code == 474) {
    countries.add('EE');
  } else if (code == 475) {
    countries.add('LV');
  } else if (code == 476) {
    countries.add('AZ');
  } else if (code == 477) {
    countries.add('LT');
  } else if (code == 478) {
    countries.add('UZ');
  } else if (code == 479) {
    countries.add('LK');
  } else if (code == 480) {
    countries.add('PH');
  } else if (code == 481) {
    countries.add('BY');
  } else if (code == 482) {
    countries.add('UA');
  } else if (code == 483) {
    countries.add('TM');
  } else if (code == 484) {
    countries.add('MD');
  } else if (code == 485) {
    countries.add('AM');
  } else if (code == 486) {
    countries.add('GE');
  } else if (code == 487) {
    countries.add('KZ');
  } else if (code == 488) {
    countries.add('TJ');
  } else if (code == 489) {
    countries.add('HK');
  } else if (490 <= code && code <= 499) {
    countries.add('JP');
  } else if (500 <= code && code <= 509) {
    countries.add('GB'); //'UK');
  } else if (520 <= code && code <= 521) {
    countries.add('GR');
  } else if (code == 528) {
    countries.add('LB');
  } else if (code == 529) {
    countries.add('CY');
  } else if (code == 530) {
    countries.add('AL');
  } else if (code == 531) {
    countries.add('MK');
  } else if (code == 535) {
    countries.add('MT');
  } else if (code == 539) {
    countries.add('IE');
  } else if (540 <= code && code <= 549) {
    countries.addAll(['BE', 'LU']);
  } else if (code == 560) {
    countries.add('PT');
  } else if (code == 569) {
    countries.add('IS');
  } else if (570 <= code && code <= 579) {
    countries.addAll(['DK']); //, 'FO', 'GL']);
  } else if (code == 590) {
    countries.add('PL');
  } else if (code == 594) {
    countries.add('RO');
  } else if (code == 599) {
    countries.add('HU');
  } else if (600 <= code && code <= 601) {
    countries.add('ZA');
  } else if (code == 603) {
    countries.add('GH');
  } else if (code == 604) {
    countries.add('SN');
  } else if (code == 608) {
    countries.add('BH');
  } else if (code == 609) {
    countries.add('MU');
  } else if (code == 611) {
    countries.add('MA');
  } else if (code == 613) {
    countries.add('DZ');
  } else if (code == 615) {
    countries.add('NG');
  } else if (code == 616) {
    countries.add('KE');
  } else if (code == 617) {
    countries.add('CM');
  } else if (code == 618) {
    countries.add('CI');
  } else if (code == 619) {
    countries.add('TN');
  } else if (code == 620) {
    countries.add('TZ');
  } else if (code == 621) {
    countries.add('SY');
  } else if (code == 622) {
    countries.add('EG');
  } else if (code == 623) {
    countries.add('BN');
  } else if (code == 624) {
    countries.add('LY');
  } else if (code == 625) {
    countries.add('JO');
  } else if (code == 626) {
    countries.add('IR');
  } else if (code == 627) {
    countries.add('KW');
  } else if (code == 628) {
    countries.add('SA');
  } else if (code == 629) {
    countries.add('AE');
  } else if (code == 630) {
    countries.add('QA');
  } else if (640 <= code && code <= 649) {
    countries.add('FI');
  } else if (690 <= code && code <= 699) {
    countries.add('CN');
  } else if (700 <= code && code <= 709) {
    countries.add('NO');
  } else if (code == 729) {
    countries.add('IL');
  } else if (730 <= code && code <= 739) {
    countries.add('SE');
  } else if (code == 740) {
    countries.add('GT');
  } else if (code == 741) {
    countries.add('SV');
  } else if (code == 742) {
    countries.add('HN');
  } else if (code == 743) {
    countries.add('NI');
  } else if (code == 744) {
    countries.add('CR');
  } else if (code == 745) {
    countries.add('PA');
  } else if (code == 746) {
    countries.add('DO');
  } else if (code == 750) {
    countries.add('MX');
  } else if (754 <= code && code <= 755) {
    countries.add('CA');
  } else if (code == 759) {
    countries.add('VE');
  } else if (760 <= code && code <= 769) {
    countries.addAll(['CH']); //, 'LI']);
  } else if (770 <= code && code <= 771) {
    countries.add('CO');
  } else if (code == 773) {
    countries.add('UY');
  } else if (code == 775) {
    countries.add('PE');
  } else if (code == 777) {
    countries.add('BO');
  } else if (778 <= code && code <= 779) {
    countries.add('AR');
  } else if (code == 780) {
    countries.add('CL');
  } else if (code == 784) {
    countries.add('PY');
  } else if (code == 786) {
    countries.add('EC');
  } else if (789 <= code && code <= 790) {
    countries.add('BR');
  } else if (800 <= code && code <= 839) {
    countries.addAll(['IT']); //, 'SM', 'VA']);
  } else if (840 <= code && code <= 849) {
    countries.addAll(['ES']); //, 'AD']);
  } else if (code == 850) {
    countries.add('CU');
  } else if (code == 858) {
    countries.add('SK');
  } else if (code == 859) {
    countries.add('CZ');
  } else if (code == 860) {
    countries.add('RS');
  } else if (code == 865) {
    countries.add('MN');
  } else if (code == 867) {
    countries.add('KP');
  } else if (868 <= code && code <= 869) {
    countries.add('TR');
  } else if (870 <= code && code <= 879) {
    countries.add('NL');
  } else if (code == 880) {
    countries.add('KR');
  } else if (code == 883) {
    countries.add('MM');
  } else if (code == 884) {
    countries.add('KH');
  } else if (code == 885) {
    countries.add('TH');
  } else if (code == 888) {
    countries.add('SG');
  } else if (code == 890) {
    countries.add('IN');
  } else if (code == 893) {
    countries.add('VN');
  } else if (code == 896) {
    countries.add('PK');
  } else if (code == 899) {
    countries.add('ID');
  } else if (900 <= code && code <= 919) {
    countries.add('AT');
  } else if (930 <= code && code <= 939) {
    countries.add('AU');
  } else if (940 <= code && code <= 949) {
    countries.add('NZ');
  } else if (code == 950) {
    // GS1 Global Office: Special applications
  } else if (code == 951) {
    // Used to issue General Manager Numbers for the EPC
    // General Identifier(GID) scheme as defined by EPC Tag Data Standard
  } else if (code == 952) {
    // Used for demonstrations and examples the GS1 system
  } else if (code == 955) {
    countries.add('MY');
  } else if (code == 958) {
    countries.add('MO');
  } else if (960 <= code && code <= 961) {
    // GS1 UK Office: GTIN-8 allocations
  } else if (962 <= code && code <= 969) {
    // GS1 Global Office: TGIN-8 allocations
  } else if (code == 977) {
    // Serial publications
  } else if (978 <= code && code <= 979) {
    // "Bookland" (ISBN) - 979-0 used for sheet music
    // ("Musicland", ISMN-13, replaces deprecated ISMN M-numbers)
  } else if (code == 980) {
    // Refund receipts
  } else if (981 <= code && code <= 984) {
    // GS1 coupon identification for common currency areas
  } else if (990 <= code && code <= 999) {
    // GS1 coupon identification
  }

  return countries;
}
